File: gm-utils/js/src/commom/utils/comum/CoordenadasMovimento.js
import IntegerParse from '../integer/IntegerParse';

export default class CoordenadasMovimento {

	xa = 0;
	xb = 0;
	ya = 0;
	yb = 0;
	inc = 0;

	constructor(xa, xb, ya, yb, inc) {

		if (inc < 1) {
			throw new Error("inc < 1");
		}
		if (xa < 0) {
			throw new Error("xa < 0");
		}
		if (xb < 0) {
			throw new Error("xb < 0");
		}
		if (ya < 0) {
			throw new Error("ya < 0");
		}
		if (yb < 0) {
			throw new Error("yb < 0");
		}

		this.inc = inc;
		this.xa = xa;
		this.xb = xb;
		this.ya = ya;
		this.yb = yb;
	}

	next() {

		if (this.xa === this.xb) {
			return this.y();
		}
		if (this.ya === this.yb) {
			return this.x();
		}

		let xc = CoordenadasMovimento.dif(this.xa, this.xb);
		let yc = CoordenadasMovimento.dif(this.xa, this.xb);

		if (xc > yc) {
			let i = IntegerParse.toInt(xc / yc);
			while (i > 0) {
				this.x();
				i--;
			}
			this.y();
		} else {
			let i = IntegerParse.toInt(yc / xc);
			while (i > 0) {
				this.y();
				i--;
			}
			this.x();
		}

		return true;

	}

	x() {
		if (this.xa === this.xb) {
			return false;
		}
		if (this.xa > this.xb) {
			this.xaDec();
		} else {
			this.xaInc();
		}
		return true;
	}

	y() {
		if (this.ya === this.yb) {
			return false;
		}
		if (this.ya > this.yb) {
			this.yaDec();
		} else {
			this.yaInc();
		}
		return true;
	}

	static dif(a, b) {
		return a > b ? a-b : b-a;
	}

	yaDec() {
		this.ya -= this.inc;
		if (this.ya < this.yb) {
			this.ya = this.yb;
		}
	}

	yaInc() {
		this.ya += this.inc;
		if (this.ya > this.yb) {
			this.ya = this.yb;
		}
	}

	xaDec() {
		this.xa -= this.inc;
		if (this.xa < this.xb) {
			this.xa = this.xb;
		}
	}

	xaInc() {
		this.xa += this.inc;
		if (this.xa > this.xb) {
			this.xa = this.xb;
		}
	}

}

File: gm-utils/src/main/java/src/commom/utils/comum/CoordenadasMovimento.java
package src.commom.utils.comum;

import src.commom.utils.integer.IntegerParse;

public class CoordenadasMovimento {

	private int xa;
	private int xb;
	private int ya;
	private int yb;
	private int inc;

	public CoordenadasMovimento(int xa, int xb, int ya, int yb, int inc) {

		if (inc < 1) {
			throw new RuntimeException("inc < 1");
		}
		if (xa < 0) {
			throw new RuntimeException("xa < 0");
		}
		if (xb < 0) {
			throw new RuntimeException("xb < 0");
		}
		if (ya < 0) {
			throw new RuntimeException("ya < 0");
		}
		if (yb < 0) {
			throw new RuntimeException("yb < 0");
		}

		this.inc = inc;
		this.xa = xa;
		this.xb = xb;
		this.ya = ya;
		this.yb = yb;
	}

	public boolean next() {

		if (xa == xb) {
			return y();
		}
		if (ya == yb) {
			return x();
		}

		int xc = dif(xa, xb);
		int yc = dif(xa, xb);

		if (xc > yc) {
			int i = IntegerParse.toInt(xc / yc);
			while (i > 0) {
				x();
				i--;
			}
			y();
		} else {
			int i = IntegerParse.toInt(yc / xc);
			while (i > 0) {
				y();
				i--;
			}
			x();
		}

		return true;

	}

	private boolean x() {
		if (xa == xb) {
			return false;
		}
		if (xa > xb) {
			xaDec();
		} else {
			xaInc();
		}
		return true;
	}

	private boolean y() {
		if (ya == yb) {
			return false;
		}
		if (ya > yb) {
			yaDec();
		} else {
			yaInc();
		}
		return true;
	}

	private static int dif(int a, int b) {
		return a > b ? a-b : b-a;
	}

	private void yaDec() {
		ya -= inc;
		if (ya < yb) {
			ya = yb;
		}
	}

	private void yaInc() {
		ya += inc;
		if (ya > yb) {
			ya = yb;
		}
	}

	private void xaDec() {
		xa -= inc;
		if (xa < xb) {
			xa = xb;
		}
	}

	private void xaInc() {
		xa += inc;
		if (xa > xb) {
			xa = xb;
		}
	}

}
